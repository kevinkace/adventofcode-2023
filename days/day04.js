export class Game {
    constructor(str) {
        this.str = str;
        this.parsed = str.match(/Card *(\d+): ([\d ]*) \| ([\d ]*)/);

        this.game = parseInt(this.parsed[1]);

        this.winners = this.parsed[2].split(/ +/).map(n => parseInt(n, 10));
        this.players = this.parsed[3].split(/ +/).map(n => parseInt(n, 10));
    }

    getWon() {
        this.won = this.won || this.players.filter(player => this.winners.includes(player));

        if (this.won.length === 0) {
            console.log(this.str);
        }

        return this.won;
    }

    getScore() {
        const won = this.getWon();

        return won.length ? Math.pow(2, won.length - 1) : 0;
    }
}

export class Games {
    constructor(str) {
        this.str = str;
        this.games = this.getGames();
    }

    getGames() {
        return this.str.split(/\r?\n/).map(game => new Game(game));
    }

    getTotalWins() {
        return this.games.map(game => game.getScore());
    }

    getTotalPoints() {
        return this.games.reduce((sum, game) => {
            return sum + game.getScore();
        }, 0)
    }
}
