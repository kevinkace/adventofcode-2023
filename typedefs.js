// Day 02

/**
 * @typedef {Object} Game
 * @property {number} game
 * @property {Round[]} rounds
 */

/**
 * @typedef {Object} Round
 * @property {number} round
 * @property {number} red
 * @property {number} green
 * @property {number} blue
 */

/**
 * @typedef {Object} Cubes
 * @property {number} red
 * @property {number} green
 * @property {number} blue
 */

// Day 3

/**
 * @typedef {Object} Game
 * @property {number} game
 * @property {number[]} winners
 * @property {number[]} players
 * @property {number[]} won
 */

export default {};
