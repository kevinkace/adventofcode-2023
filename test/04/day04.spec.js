import { readFile }      from "node:fs/promises";
import { fileURLToPath } from "node:url";
import path              from "node:path";

import { expect, test, describe } from "vitest";

import { Game, Games } from "../../days/day04.js";

import typedefs from "../../typedefs.js";

const __filename = fileURLToPath(import.meta.url);
const __dirname  = path.dirname(__filename);

const sample = await readFile(path.join(__dirname, "sample.txt"), "utf8");
const input  = await readFile(path.join(__dirname, "input.txt"), "utf8");

const sampleGame = "Card 1: 41 48 83 86 17 | 83 86  6 31 17  9 48 53";


describe("day 04", () => {
    test("can create new Game", () => {
        expect(new Game(sampleGame)).toMatchSnapshot();
    });

    test("can get game wins", () => {
        const game = new Game(sampleGame);

        expect(game.getWon()).toMatchSnapshot();
    });

    test("can get game score", () => {
        const game = new Game(sampleGame);

        expect(game.getScore()).toEqual(8);
    });

    test("can create new Games", () => {
        expect(new Games(sample)).toMatchSnapshot();
    });

    test("can get total wins", () => {
        const games = new Games(sample);

        expect(games.getTotalWins()).toMatchSnapshot();
    });

    test("can get total points", () => {
        const games = new Games(sample);

        expect(games.getTotalPoints()).toEqual(13);
    });

    test("answer p1", () => {
        const games = new Games(input);

        const points = games.getTotalPoints();

        console.log(points);
        expect(points).toBeLessThan(26478)
    });
});
