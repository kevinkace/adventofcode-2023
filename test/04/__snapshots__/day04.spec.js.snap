// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`day 04 > can create new Game 1`] = `
Game {
  "game": 1,
  "parsed": [
    "Card 1: 41 48 83 86 17 | 83 86  6 31 17  9 48 53",
    "1",
    "41 48 83 86 17",
    "83 86  6 31 17  9 48 53",
  ],
  "players": [
    83,
    86,
    6,
    31,
    17,
    9,
    48,
    53,
  ],
  "str": "Card 1: 41 48 83 86 17 | 83 86  6 31 17  9 48 53",
  "winners": [
    41,
    48,
    83,
    86,
    17,
  ],
}
`;

exports[`day 04 > can create new Games 1`] = `
Games {
  "games": [
    Game {
      "game": 1,
      "parsed": [
        "Card 1: 41 48 83 86 17 | 83 86  6 31 17  9 48 53",
        "1",
        "41 48 83 86 17",
        "83 86  6 31 17  9 48 53",
      ],
      "players": [
        83,
        86,
        6,
        31,
        17,
        9,
        48,
        53,
      ],
      "str": "Card 1: 41 48 83 86 17 | 83 86  6 31 17  9 48 53",
      "winners": [
        41,
        48,
        83,
        86,
        17,
      ],
    },
    Game {
      "game": 2,
      "parsed": [
        "Card 2: 13 32 20 16 61 | 61 30 68 82 17 32 24 19",
        "2",
        "13 32 20 16 61",
        "61 30 68 82 17 32 24 19",
      ],
      "players": [
        61,
        30,
        68,
        82,
        17,
        32,
        24,
        19,
      ],
      "str": "Card 2: 13 32 20 16 61 | 61 30 68 82 17 32 24 19",
      "winners": [
        13,
        32,
        20,
        16,
        61,
      ],
    },
    Game {
      "game": 3,
      "parsed": [
        "Card 3:  1 21 53 59 44 | 69 82 63 72 16 21 14  1",
        "3",
        " 1 21 53 59 44",
        "69 82 63 72 16 21 14  1",
      ],
      "players": [
        69,
        82,
        63,
        72,
        16,
        21,
        14,
        1,
      ],
      "str": "Card 3:  1 21 53 59 44 | 69 82 63 72 16 21 14  1",
      "winners": [
        NaN,
        1,
        21,
        53,
        59,
        44,
      ],
    },
    Game {
      "game": 4,
      "parsed": [
        "Card 4: 41 92 73 84 69 | 59 84 76 51 58  5 54 83",
        "4",
        "41 92 73 84 69",
        "59 84 76 51 58  5 54 83",
      ],
      "players": [
        59,
        84,
        76,
        51,
        58,
        5,
        54,
        83,
      ],
      "str": "Card 4: 41 92 73 84 69 | 59 84 76 51 58  5 54 83",
      "winners": [
        41,
        92,
        73,
        84,
        69,
      ],
    },
    Game {
      "game": 5,
      "parsed": [
        "Card 5: 87 83 26 28 32 | 88 30 70 12 93 22 82 36",
        "5",
        "87 83 26 28 32",
        "88 30 70 12 93 22 82 36",
      ],
      "players": [
        88,
        30,
        70,
        12,
        93,
        22,
        82,
        36,
      ],
      "str": "Card 5: 87 83 26 28 32 | 88 30 70 12 93 22 82 36",
      "winners": [
        87,
        83,
        26,
        28,
        32,
      ],
    },
    Game {
      "game": 6,
      "parsed": [
        "Card 6: 31 18 13 56 72 | 74 77 10 23 35 67 36 11",
        "6",
        "31 18 13 56 72",
        "74 77 10 23 35 67 36 11",
      ],
      "players": [
        74,
        77,
        10,
        23,
        35,
        67,
        36,
        11,
      ],
      "str": "Card 6: 31 18 13 56 72 | 74 77 10 23 35 67 36 11",
      "winners": [
        31,
        18,
        13,
        56,
        72,
      ],
    },
  ],
  "str": "Card 1: 41 48 83 86 17 | 83 86  6 31 17  9 48 53
Card 2: 13 32 20 16 61 | 61 30 68 82 17 32 24 19
Card 3:  1 21 53 59 44 | 69 82 63 72 16 21 14  1
Card 4: 41 92 73 84 69 | 59 84 76 51 58  5 54 83
Card 5: 87 83 26 28 32 | 88 30 70 12 93 22 82 36
Card 6: 31 18 13 56 72 | 74 77 10 23 35 67 36 11",
}
`;

exports[`day 04 > can get game wins 1`] = `
[
  83,
  86,
  17,
  48,
]
`;

exports[`day 04 > can get total wins 1`] = `
[
  8,
  2,
  2,
  1,
  0,
  0,
]
`;
